src/menu.rs

```diff
@@ -0,0 +1,525 @@
+            hotkey: None,
+            tool_tip: None,
+            selected: false,
+            enabled: true,
+            platform_id: MenuItemId::PLACEHOLDER,
+        }
+    }
+
+    /// A builder method that adds a hotkey for this item.
+    ///
+    /// # Example
+    ///
+    /// ```
+    /// # use druid::{LocalizedString, MenuDesc, Selector, SysMods};
+    /// # use druid::menu::MenuItem;
+    ///
+    /// let item = MenuItem::new(LocalizedString::new("My Menu Item"), Selector::new("My Selector"))
+    ///     .hotkey(SysMods::Cmd, "m");
+    ///
+    /// # // hide the type param in or example code by letting it be inferred here
+    /// # MenuDesc::<u32>::empty().append(item);
+    /// ```
+    pub fn hotkey(mut self, mods: impl Into<Option<RawMods>>, key: impl Into<KeyCompare>) -> Self {
+        self.hotkey = Some(HotKey::new(mods, key));
+        self
+    }
+
+    /// Disable this menu item.
+    pub fn disabled(mut self) -> Self {
+        self.enabled = false;
+        self
+    }
+
+    /// Disable this menu item if the provided predicate is true.
+    pub fn disabled_if(mut self, mut p: impl FnMut() -> bool) -> Self {
+        if p() {
+            self.enabled = false;
+        }
+        self
+    }
+
+    /// Mark this menu item as selected. This will usually be indicated by
+    /// a checkmark.
+    pub fn selected(mut self) -> Self {
+        self.selected = true;
+        self
+    }
+
+    /// Mark this item as selected, if the provided predicate is true.
+    pub fn selected_if(mut self, mut p: impl FnMut() -> bool) -> Self {
+        if p() {
+            self.selected = true;
+        }
+        self
+    }
+}
+
+impl<T: Data> MenuDesc<T> {
+    /// Create a new, empty menu.
+    pub fn empty() -> Self {
+        Self::new(LocalizedString::new(""))
+    }
+
+    /// Create a new menu with the given title.
+    pub fn new(title: LocalizedString<T>) -> Self {
+        let item = MenuItem::new(title, Selector::NOOP);
+        MenuDesc {
+            item,
+            items: Vec::new(),
+        }
+    }
+
+    /// Given a function that produces an iterator, appends that iterator's
+    /// items to this menu.
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// use druid::{Command, LocalizedString, MenuDesc, Selector};
+    /// use druid::menu::MenuItem;
+    ///
+    /// let num_items: usize = 4;
+    /// const MENU_COUNT_ACTION: Selector = Selector::new("menu-count-action");
+    ///
+    /// let my_menu: MenuDesc<u32> = MenuDesc::empty()
+    ///     .append_iter(|| (0..num_items).map(|i| {
+    ///         MenuItem::new(
+    ///             LocalizedString::new("hello-counter").with_arg("count", move |_, _| i.into()),
+    ///             Command::new(MENU_COUNT_ACTION, i),
+    ///        )
+    ///     })
+    /// );
+    ///
+    /// assert_eq!(my_menu.len(), 4);
+    /// ```
+    pub fn append_iter<I: Iterator<Item = MenuItem<T>>>(mut self, f: impl FnOnce() -> I) -> Self {
```

Comment:
`extend` is part of the `Extend` trait, so I'm reluctant to implement it directly without us being able to use that trait.

I did try and have this take an iterator directly (instead of a closure that produces an iterator) but there was some problem I ran into. 😕